src/v/cluster/metadata_cache.h

```diff
@@ -72,8 +73,24 @@ class metadata_cache {
     ///\brief Returns metadata of single topic.
     ///
     /// If topic does not exists it returns an empty optional
-    std::optional<model::topic_metadata> get_topic_metadata(
+    std::optional<model::topic_metadata> get_model_topic_metadata(
       model::topic_namespace_view, with_leaders = with_leaders::yes) const;
+    ///\brief Returns metadata of single topic.
+    ///
+    /// If topic does not exists it returns an empty optional
+    std::optional<cluster::topic_metadata>
+      get_topic_metadata(model::topic_namespace_view) const;
+
+    ///\brief Returns reference to metadata of single topic.
+    ///
+    /// If topic does not exists it returns an empty optional
+    ///
+    /// IMPORTANT: remember not to use the reference when its lifetime has to
+    /// span across multiple scheduling point. Reference returning method is
+    /// provided not to copy metadata object when used by synchronous parts of
+    /// code
+    std::optional<std::reference_wrapper<const cluster::topic_metadata>>
```

Comment:
Nice comment, thanks!

Reply:
@mmaslankaprv thanks for the comment.

It might be worthwhile to consider making this harder to use incorrectly. Perhaps by wrapping in a type that can't be copied or moved, so that it doesn't escape the scope as easily. The issue is that it really _looks_ like it has value semantics at the call site:

        auto metadata = _topics.local().get_topic_metadata_ref(tp_ns);

Reply:
good point, i will consider wrapping it in a noncopyable nonmovable type.

Reply:
nbd. just thoughts in passing reading old PRs